cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepository.java

```diff
@@ -0,0 +1,245 @@
+import io.cdap.cdap.security.impersonation.Impersonator;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import javax.annotation.Nullable;
+
+/**
+ * RemoteArtifactRepository provides a remote implementation of ArtifactRepository
+ */
+public class RemoteArtifactRepository implements ArtifactRepository {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteArtifactRepository.class);
+  private final ArtifactRepositoryReader artifactRepositoryReader;
+  private final ArtifactClassLoaderFactory artifactClassLoaderFactory;
+  private final ArtifactInspector artifactInspector;
+  private final Set<File> systemArtifactDirs;
+  private final ArtifactConfigReader configReader;
+  private final MetadataServiceClient metadataServiceClient;
+  private final Impersonator impersonator;
+
+  @VisibleForTesting
+  @Inject
+  public RemoteArtifactRepository(CConfiguration cConf, ArtifactRepositoryReader artifactRepositoryReader,
+                                  MetadataServiceClient metadataServiceClient,
+                                  ProgramRunnerFactory programRunnerFactory,
+                                  Impersonator impersonator) {
+    this.artifactRepositoryReader = artifactRepositoryReader;
+    this.artifactClassLoaderFactory = new ArtifactClassLoaderFactory(cConf, programRunnerFactory);
+    this.artifactInspector = new ArtifactInspector(cConf, artifactClassLoaderFactory);
+    this.systemArtifactDirs = new HashSet<>();
+    String systemArtifactsDir = cConf.get(Constants.AppFabric.SYSTEM_ARTIFACTS_DIR);
+    if (!Strings.isNullOrEmpty(systemArtifactsDir)) {
+      String sparkDirStr = SparkCompatReader.get(cConf).getCompat();
+
+      for (String dir : systemArtifactsDir.split(";")) {
+        File file = new File(dir);
+        if (!file.isDirectory()) {
+          LOG.warn("Ignoring {} because it is not a directory.", file);
+          continue;
+        }
+        systemArtifactDirs.add(file);
+        // Also look in the relevant spark compat directory for spark version specific artifacts.
+        File sparkDir = new File(file, sparkDirStr);
+        if (file.isDirectory()) {
+          systemArtifactDirs.add(sparkDir);
+        }
+      }
+    }
+    this.configReader = new ArtifactConfigReader();
+    this.metadataServiceClient = metadataServiceClient;
+    this.impersonator = impersonator;
+  }
+
+  @Override
+  public CloseableClassLoader createArtifactClassLoader(Location artifactLocation,
+                                                        EntityImpersonator entityImpersonator) throws IOException {
+    return artifactClassLoaderFactory.createClassLoader(ImmutableList.of(artifactLocation).iterator(),
```

Comment:
Why not just call `createClassLoader(artifactLocation, entityImpersonator);` instead?

Reply:
The method that accepts a single location (instead of an iterator) is private: 

```
private CloseableClassLoader createClassLoader(final Location artifactLocation,  EntityImpersonator entityImpersonator) throws IOException {
```

I've removed the private modifier so its visible to other classes now, let me know if it should stay private